src/bindings/python/flux/kvs.py

```diff
@@ -0,0 +1,216 @@
+from _kvs import ffi, lib
+import flux
+from flux.wrapper import Wrapper, WrapperPimpl, WrapperBase
+import flux.json_c as json_c
+import json
+import collections
+import contextlib
+import errno
+
+raw = Wrapper(ffi, lib, prefixes=['kvs','kvs_'])
+ 
+
+class KVSHandle(Wrapper):
+  def __init__(self, flux_handle=None):
+    if flux_handle is None:
+      raise ValueError("flux_handle must be a valid Flux object")
+    super(self.__class__, self).__init__(ffi, lib, handle=flux_handle,
+                                     match=ffi.typeof(lib.flux_open).result,
+                                     prefixes=[
+                                       'kvs_',
+                                       ],
+                                     )
+
+class KVSDir(WrapperPimpl,collections.MutableMapping):
+  class InnerWrapper(Wrapper):
+    def __init__(self, flux_handle=None, path='.', handle=None):
+      self.handle = None
+      global raw
+      if flux_handle is None and handle is None:
+        raise ValueError("flux_handle must be a valid Flux object or handle must be a valid kvsdir cdata pointer")
+      if isinstance(flux_handle, WrapperBase):
+        flux_handle = flux_handle.handle
+      if handle is None:
+        d = ffi.new("kvsdir_t [1]")
+        raw.kvs_get_dir(flux_handle, d, path)
+        handle = d[0]
+        if handle is None or handle == ffi.NULL:
+          raise IOError(errno.ENOENT, os.strerror(errno.ENOENT))
+
+      super(self.__class__, self).__init__(ffi, lib, handle=handle,
+                                       match=ffi.typeof('kvsdir_t'),
+                                       prefixes=[
+                                         'kvsdir_',
+                                         ],
+                                       )
+    def __del__(self):
+      if self.handle is not None:
+        raw.kvsdir_destroy(self.handle)
+
+  def __init__(self, flux_handle=None, path='.', handle=None):
+      self.fh = flux_handle
+      self.path = path
+      if flux_handle is None and handle is None:
+        raise ValueError("flux_handle must be a valid Flux object or handle must be a valid kvsdir cdata pointer")
+      self.pimpl = self.InnerWrapper(flux_handle, path, handle)
+
+  def commit(self):
+    global raw
+    ret = raw.kvs_commit(self.fh.handle)
+    if ret < 0:
+      raise EnvironmentError(ret)
+
+  def key_at(self, key):
+    c_str = self.pimpl.key_at(key)
+    p_str = ffi.string(c_str)
+    lib.free(c_str)
+    return p_str
+
+  def __getitem__(self, key):
+    try:
+      j = json_c.Jobj()
+      self.pimpl.get(key, j.get_as_dptr())
+      return json.loads(j.as_str())
+    except EnvironmentError as err:
+      if err == errno.EISDIR:
```

Comment:
So this doesn't work as expected for kvs directories, but the below should fix it:
```python
if err.errno == errno.EISDIR:
```

Unfortunately, I'm still getting Key/Env Errors on non-integer/non-directory values. I tried iterating over the values in the lwj directory (lwj.1.nnodes, lwj.1.create-time, etc) and I get errors for any non-integer value.  Most give me KeyErrors, but one (and sub-directory) gives me an EnvError which passes through b/c the errno != EISDIR. The minimal test case along with the output is below:
```python
    with kvs.get_dir(h, "lwj.1") as lwj_dir:
        for prop_name in lwj_dir:
                try:
                    print "\t{} {}".format(prop_name, lwj_dir[prop_name])
                except EnvironmentError as err:
                    print "EnvError: {}, Key: {}".format(err, prop_name)
                except KeyError as err:
                    print "KeyError:", err
```
Note: All of the errors below have a errno of 9: Bad file descriptor
```
        nnodes 8
        ntasks 128
KeyError: 'create-time'
KeyError: 'user'
KeyError: 'jobname'
KeyError: 'account'
KeyError: 'submit_time'
KeyError: 'execution_time'
KeyError: 'time_limit'
        ncpus 128
        io_rate 2
KeyError: 'allocated-time'
EnvError: [Errno 9] Bad file descriptor, Key: rank
KeyError: 'runrequest-time'
KeyError: 'starting_time'
KeyError: 'running_time'
KeyError: 'state'
KeyError: 'complete_time'
KeyError: 'io_time'
```